Add populator flow (#367)
kubev2v/forklift-documentation: documentation/modules/virt-migration-workflow.adoc

The workflow describes the following steps:

> *Warm Migration and remote clusters:*

Reviewer: ```suggestion
*Warm Migration or migration to a remote OCP cluster:*
```

+
*After the VM disks are transferred:*

. The `Migration Controller` service creates a `conversion` pod with the PVCs attached to it when importing from VMWare.
+
The `conversion` pod runs `virt-v2v`, which installs and configures device drivers on the PVCs of the target VM.
+
. The `Migration Controller` service creates a `VirtualMachine` CR for each source virtual machine (VM), connected to the PVCs.

> . When the target VM is powered on, the `KubeVirt Controller` service creates a `virt-launcher` pod and a `VirtualMachineInstance` CR.

Reviewer: we start the vm automatically on the destination when the users asks us to so for this particular case I'd keep it, otherwise what happens after the migration is finished is out of scope. so instead of "When the target VM is powered on" I'd write (but need to double check me on this): "when the option to restore the power state of the VM is checked and the VM was running on the source environment, the `Migration Contoller` powers on the VM .."

Author: the user can't ask for it. if the source vm is powered on, it will automatically power-on on destination

Reviewer: indeed, I had a vague memory of being able to choose - not sure why..

+
The `virt-launcher` pod runs `QEMU-KVM` with the PVCs attached as VM disks.

> *Cold migration:*

Author: This is relevant for ovirt/openstack. But, in VMware it's different now (also the above flow). Shall I add it?

Reviewer 2: @ahadas - please can we ask for your opinion?

Reviewer: yes please. no need to get too much into the details - just to indicate that in that specific flow the conversion pod also imports the disks rather than CDI

Author: updated, @nyoxi please confirm.

> . When you create a `Migration` custom resource (CR) to run a migration plan, the `Migration Controller` service creates a `PersistentVolumeClaim` CR and a `OvirtVolumePopulator`/`OpenstackVolumePopulator` CR for each source VM disk.

Reviewer: when the source is oVirt/RHV and it's a cold-migration to the local OCP cluster or when the source is Openstack

Author: the titles aren't enough?

Reviewer: still relevant? I see you've changed it

+
*For each VM disk:*

> . The `Populator Controller` service creates a temporarily persistent volume claim (PVC).

Reviewer: need to also add the `Populator Controller` to the "MTV services" section then

. If the `StorageClass` has a dynamic provisioner, the persistent volume (PV) is dynamically provisioned by the `StorageClass` provisioner.

> . The `Migration Controller` service creates a dummy pod to bound the PVC.

Reviewer: we can identify it by its name, right? let's write that
and "to bound" -> "to bind"

. The `Populator Controller` service creates a `populator` pod.

> . The `populator` pod streams the VM disk to the PV.

Reviewer: ```suggestion
. The `populator` pod transfers the disk data to the PV.
```

+
*After the VM disks are transferred:*
. The temporarily PVC is deleted, and the initial PVC will point to the PV with the data.
. The `Migration Controller` service creates a `VirtualMachine` CR for each source virtual machine (VM), connected to the PVCs.
. If the VM was running on the source environment, the `Migration Controller` powers on the VM, the `KubeVirt Controller` service creates a `virt-launcher` pod and a `VirtualMachineInstance` CR.
+
The `virt-launcher` pod runs `QEMU-KVM` with the PVCs attached as VM disks.

*Cold migration (VMWare):*

. When you create a `Migration` custom resource (CR) to run a migration plan, the `Migration Controller` service creates a `DataVolume` CR for each source VM disk.
+
*For each VM disk:*

. The `Containerized Data Importer (CDI) Controller` service creates a blank persistent volume claim (PVC) based on the parameters specified in the `DataVolume` CR.  
. If the `StorageClass` has a dynamic provisioner, the persistent volume (PV) is dynamically provisioned by the `StorageClass` provisioner.
. The `Migration Controller` service creates a dummy pod to bind the PVC, the pod contains `pvcinit` in his name.

> . The `Migration Controller` service creates a `conversion` pod.

Reviewer 3: Ths does not really fit into the "For each VM disk" section. Same for the `pvcinit` pod (here and elsewhere). Although I am not sure if it's worth separating to a new section.

Author: i tried to pull it out (easily in vmware hard in ovirt because the flow). making it without separating to a new section will make the section much harder to read in my opinion

Reviewer 3: Ok, let's leave it as it is now.

+
*After the VM disks are transferred:*

. The `Migration Controller` service creates a `conversion` pod with the PVCs attached to it when importing from VMWare.
+
The `conversion` pod runs `virt-v2v`, which installs and configures device drivers on the PVCs of the target VM.
+
. The `Migration Controller` service creates a `VirtualMachine` CR for each source virtual machine (VM), connected to the PVCs.

> . If the VM was running on the source environment, the `Migration Controller` powers on the VM, the `KubeVirt Controller` service creates a `virt-launcher` pod and a `VirtualMachineInstance` CR.

Reviewer 4: consider: ran

Reviewer 2: resolved

+
The `virt-launcher` pod runs `QEMU-KVM` with the PVCs attached as VM disks.

> *Cold migration (oVirt/OpenStack):*

Reviewer: ```suggestion
*Cold migration from {rhv-short} or {osp} to the local OCP cluster:*
```

> . When you create a `Migration` custom resource (CR) to run a migration plan, the `Migration Controller` service creates a `PersistentVolumeClaim` CR and a `OvirtVolumePopulator` CR for each source VM disk when the source is oVirt/RHV, and it's a cold-migration to the local OCP cluster or `OpenstackVolumePopulator` CR for each source VM disk when the source is Openstack. 

Reviewer: need to replace oVirt/RHV with the variable we have for that

> . When you create a `Migration` custom resource (CR) to run a migration plan, the `Migration Controller` service creates a `PersistentVolumeClaim` CR and a `OvirtVolumePopulator` CR for each source VM disk when the source is {rhv-short}, and it's a cold-migration to the local OCP cluster or `OpenstackVolumePopulator` CR for each source VM disk when the source is Openstack.

Reviewer: ```suggestion
. When you create a `Migration` custom resource (CR) to run a migration plan, the `Migration Controller` service creates for each source VM disk a `PersistentVolumeClaim` CR, and a `OvirtVolumePopulator` when the source is {rhv-short}, or a `OpenstackVolumePopulator` CR  when the source is {osp}.
```

+
*For each VM disk:*

. The `Populator Controller` service creates a temporarily persistent volume claim (PVC).
. If the `StorageClass` has a dynamic provisioner, the persistent volume (PV) is dynamically provisioned by the `StorageClass` provisioner.

> - The `Migration Controller` service creates a dummy pod to bind *all PVCs*, the name of the pod contains `pvcinit`.

Reviewer 4: Consider: Two sentences.  "The ... creates a single dummy pod that bind *all* PVCs. The name of the pod includes `pvcinit`.  If "all PVCs" is a UI name, bold it.

> - The `Migration Controller` service creates a dummy pod to bind *all PVCs*, the pod contains `pvcinit` in his name.

Reviewer: ```suggestion
- The `Migration Controller` service creates a dummy pod to bind *all PVCs*, the name of the pod contains `pvcinit`.
```

. The `Populator Controller` service creates a `populator` pod.
. The `populator` pod transfers the disk data to the PV.
+
*After the VM disks are transferred:*

> . The temporarily PVC is deleted, and the initial PVC will point to the PV with the data.

Reviewer 3: ```suggestion
. The temporary PVC is deleted, and the initial PVC will point to the PV with the data.
```

> . The temporary PVC is deleted, and the initial PVC will point to the PV with the data.

Reviewer 4: points to

. The `Migration Controller` service creates a `VirtualMachine` CR for each source virtual machine (VM), connected to the PVCs.
. If the VM was running on the source environment, the `Migration Controller` powers on the VM, the `KubeVirt Controller` service creates a `virt-launcher` pod and a `VirtualMachineInstance` CR.

> +

Reviewer 2: resolved

The `virt-launcher` pod runs `QEMU-KVM` with the PVCs attached as VM disks.

> *Cold migration (VMWare):*

Reviewer: ```suggestion
*Cold migration from VMWare to the local OCP cluster:*
```

. When you create a `Migration` custom resource (CR) to run a migration plan, the `Migration Controller` service creates a `DataVolume` CR for each source VM disk.
+
*For each VM disk:*

. The `Containerized Data Importer (CDI) Controller` service creates a blank persistent volume claim (PVC) based on the parameters specified in the `DataVolume` CR.  
. If the `StorageClass` has a dynamic provisioner, the persistent volume (PV) is dynamically provisioned by the `StorageClass` provisioner.

*For all VM disks:*

> . The `Migration Controller` service creates a dummy pod to bind *all PVCs*, the name of the pod contains `pvcinit`.

Reviewer 4: Consider: Two sentences. "The ... creates a single dummy pod that bind all PVCs. The name of the pod includes pvcinit. If "all PVCs" is a UI name, bold it.

> . The `Migration Controller` service creates a dummy pod to bind *all PVCs*, the pod contains `pvcinit` in his name.

Reviewer: ```suggestion
. The `Migration Controller` service creates a dummy pod to bind *all PVCs*, the name of the pod contains `pvcinit`.
```

. The `Migration Controller` service creates a `conversion` pod for *all PVCs*.

> . The `conversion` pod runs `virt-v2v`, which transfers the disks data to the PVs and converts the disk to the KVM hypervisor.

Reviewer 2: Please can we change "**transfers the disks data to...**" to "**transfers the disk's data to...**" 
Thanks

> . The `conversion` pod runs `virt-v2v`, which transfers the disk's data to the PVs and converts the disk to the KVM hypervisor.

Reviewer: I think we better switch between the transfer and the conversion phases to be more accurate and go with the plural form here:
```suggestion
. The `conversion` pod runs `virt-v2v`, which converts the VM to the KVM hypervisor and transfers the disks' data to their corresponding PVs.
```

> . The `conversion` pod runs `virt-v2v`, which transfers the disks data to the PVs, installs and configures device drivers on the PVCs of the target VM.

Reviewer: ```suggestion
. The `conversion` pod runs `virt-v2v`, which transfers the disks data to the PVs and converts the disk to the KVM hypervisor.
```
